### PR TITLE
Write a new .proto file loader that will work as WireRun prescribes

### PR DESCRIPTION
**Pull request by swankjesse**
It's more flexible in how inputs are provided - fully-qualified paths,
directories, or .zips. And it validates that the parsed paths are in
the expected directories.

The Root class keeps track of a user-provided directory where .proto
files are loaded.

The NewSchemaLoader loads the user's provided sources, and the protos
they import.

### CODE REVIEW

File: wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        val sourceFs = FileSystems.newFileSystem(this, javaClass.classLoader)
+        closer.register(sourceFs)
+        sourceFs.rootDirectories.map { inZipDirectory -> DirectoryRoot(this, inZipDirectory) }
+      } catch (e: ProviderNotFoundException) {
+        throw IllegalArgumentException(
+            "expected a directory, archive (.zip / .jar / etc.), or .proto: $this")
+      }
+    }
+  }
+}
```
Comment (jrodbx) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt]: nice

File: wire-compiler/src/test/java/com/squareup/wire/schema/NewSchemaLoaderTest.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        |import "squareup/polygons/triangle.proto";
+        |message Circle {
+        |}
+        """.trimMargin(),
+        "squareup/curves/oval.proto" to """
+        |syntax = "proto2";
+        |package squareup.curves;
+        |message Oval {
+        |}
+        """.trimMargin())
```
Comment (jrodbx) [inline wire-compiler/src/test/java/com/squareup/wire/schema/NewSchemaLoaderTest.kt]: 👌

File: wire-compiler/src/test/java/com/squareup/wire/schema/NewSchemaLoaderTest.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    val sourcePath = listOf("colors/src/main/proto")
+    val exception = assertFailsWith<IllegalArgumentException> {
+      NewSchemaLoader(fs, sourcePath).use { it.load() }
+    }
+    assertThat(exception).hasMessage("expected colors/src/main/proto/squareup/shapes/blue.proto " +
+        "to have a path ending with squareup/colors/blue.proto")
+  }
```
Comment (jrodbx) [inline wire-compiler/src/test/java/com/squareup/wire/schema/NewSchemaLoaderTest.kt]: good edge case

File: wire-compiler/src/test/java/com/squareup/wire/schema/RootTest.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import java.nio.file.FileSystem
+import java.nio.file.Files
+import java.util.zip.ZipEntry
+import java.util.zip.ZipOutputStream
+import kotlin.test.assertFailsWith
+import kotlin.text.Charsets.UTF_8
+
+class RootTest {
+  private val fs = Jimfs.newFileSystem(Configuration.unix())
+  private val closer = Closer.create()
```
Comment (jrodbx) [inline wire-compiler/src/test/java/com/squareup/wire/schema/RootTest.kt]: nit: explicit setup in @Before?

File: wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+internal fun Path.roots(closer: Closer): List<Root> {
+  return when {
+    Files.isDirectory(this) -> listOf(DirectoryRoot(this, this))
+
+    endsWithDotProto() -> listOf(StandaloneFileRoot(this))
+
+    // Handle a .zip or .jar file by adding all .proto files within.
+    else -> {
+      try {
+        val sourceFs = FileSystems.newFileSystem(this, javaClass.classLoader)
```
Comment (jrodbx) [inline wire-compiler/src/main/java/com/squareup/wire/schema/Root.kt]: TIL

**Comment by jrodbx** [comment]
LGTM!

File: wire-compiler/src/test/java/com/squareup/wire/schema/RootTest.kt (hunk 4 above)
Comment (swankjesse) [reply wire-compiler/src/test/java/com/squareup/wire/schema/RootTest.kt]: I think it’s less code as-is? But I agree it’s asymmetric
